File: app/src/test/java/bj/discogsbrowser/order/OrderControllerTest.java
package bj.discogsbrowser.order;

import android.content.Context;

import com.airbnb.epoxy.EpoxyModel;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import java.util.List;

import bj.discogsbrowser.utils.analytics.AnalyticsTracker;
import bj.discogsbrowser.utils.ImageViewAnimator;

import static android.os.Build.VERSION_CODES.LOLLIPOP;
import static junit.framework.Assert.assertEquals;

/**
 * Created by Josh Laird on 10/05/2017.
 */
@RunWith(RobolectricTestRunner.class)
@Config(sdk = LOLLIPOP, manifest = Config.NONE)
public class OrderControllerTest
{
    private OrderController controller;
    @Mock Context context;
    @Mock OrderContract.View view;
    @Mock ImageViewAnimator imageViewAnimator;
    @Mock AnalyticsTracker tracker;
    private OrderFactory orderFactory = new OrderFactory();

    @Before
    public void setup()
    {
        MockitoAnnotations.initMocks(this);
        controller = new OrderController(context, view, imageViewAnimator, tracker);
        controller.requestModelBuild();
    }

    @Test
    public void initialLoadingState_correct()
    {
        List<EpoxyModel<?>> copyOfModels = controller.getAdapter().getCopyOfModels();
        assertEquals(copyOfModels.get(0).getClass().getSimpleName(), "LoadingModel_");
        assertEquals(copyOfModels.size(), 1);
    }

    @Test
    public void errorThenRetry_displaysErrorThenValidData()
    {
        controller.setError(true);
        List<EpoxyModel<?>> copyOfModels = controller.getAdapter().getCopyOfModels();
        assertEquals(copyOfModels.get(0).getClass().getSimpleName(), "RetryModel_");
        assertEquals(copyOfModels.size(), 1);

        controller.setLoadingOrder(true);
        copyOfModels = controller.getAdapter().getCopyOfModels();
        assertEquals(copyOfModels.get(0).getClass().getSimpleName(), "LoadingModel_");
        assertEquals(copyOfModels.size(), 1);

        controller.setOrderDetails(orderFactory.getOneItemOrder());
        copyOfModels = controller.getAdapter().getCopyOfModels();
        assertEquals(copyOfModels.get(0).getClass().getSimpleName(), "BuyerModel_");
        assertEquals(copyOfModels.get(1).getClass().getSimpleName(), "DividerModel_");
        assertEquals(copyOfModels.get(2).getClass().getSimpleName(), "OrderReleaseModel_");
        assertEquals(copyOfModels.get(3).getClass().getSimpleName(), "DividerModel_");
        assertEquals(copyOfModels.get(4).getClass().getSimpleName(), "TotalModel_");
        assertEquals(copyOfModels.size(), 5);
    }

    @Test
    public void fourItems_displaysFourItems()
    {
        controller.setOrderDetails(orderFactory.getFourItemsOrder());
        List<EpoxyModel<?>> copyOfModels = controller.getAdapter().getCopyOfModels();
        assertEquals(copyOfModels.get(0).getClass().getSimpleName(), "BuyerModel_");
        assertEquals(copyOfModels.get(1).getClass().getSimpleName(), "DividerModel_");
        assertEquals(copyOfModels.get(2).getClass().getSimpleName(), "OrderReleaseModel_");
        assertEquals(copyOfModels.get(3).getClass().getSimpleName(), "OrderReleaseModel_");
        assertEquals(copyOfModels.get(4).getClass().getSimpleName(), "OrderReleaseModel_");
        assertEquals(copyOfModels.get(5).getClass().getSimpleName(), "OrderReleaseModel_");
        assertEquals(copyOfModels.get(6).getClass().getSimpleName(), "DividerModel_");
        assertEquals(copyOfModels.get(7).getClass().getSimpleName(), "TotalModel_");
        assertEquals(copyOfModels.size(), 8);
    }
}

File: app/src/test/java/bj/vinylbrowser/order/OrderControllerTest.kt
package bj.vinylbrowser.order

import android.content.Context
import android.os.Build.VERSION_CODES.LOLLIPOP
import bj.vinylbrowser.utils.ImageViewAnimator
import bj.vinylbrowser.utils.analytics.AnalyticsTracker
import com.nhaarman.mockito_kotlin.mock
import junit.framework.Assert.assertEquals
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.MockitoAnnotations
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

/**
 * Created by Josh Laird on 22/05/2017.
 */
@RunWith(RobolectricTestRunner::class)
@Config(sdk = intArrayOf(LOLLIPOP), manifest = Config.NONE)
class OrderControllerTest {
    lateinit var controller: OrderController
    val context: Context = mock()
    val view: OrderContract.View  = mock()
    val imageViewAnimator: ImageViewAnimator = mock()
    val tracker: AnalyticsTracker = mock()

    @Before
    fun setup() {
        MockitoAnnotations.initMocks(this)
        controller = OrderController(context, view, imageViewAnimator, tracker)
        controller.requestModelBuild()
    }

    @Test
    fun initialLoadingState_correct() {
        val copyOfModels = controller.adapter.copyOfModels
        assertEquals(copyOfModels[0].javaClass.simpleName, "LoadingModel_")
        assertEquals(copyOfModels.size, 1)
    }

    @Test
    fun errorThenRetry_displaysErrorThenValidData() {
        controller.setError(true)
        var copyOfModels = controller.adapter.copyOfModels
        assertEquals(copyOfModels[0].javaClass.simpleName, "RetryModel_")
        assertEquals(copyOfModels.size, 1)

        controller.setLoadingOrder(true)
        copyOfModels = controller.adapter.copyOfModels
        assertEquals(copyOfModels[0].javaClass.simpleName, "LoadingModel_")
        assertEquals(copyOfModels.size, 1)

        controller.setOrderDetails(OrderFactory.buildOneOrderWithItems(4))
        copyOfModels = controller.adapter.copyOfModels
        assertEquals(copyOfModels[0].javaClass.simpleName, "BuyerModel_")
        assertEquals(copyOfModels[1].javaClass.simpleName, "DividerModel_")
        assertEquals(copyOfModels[2].javaClass.simpleName, "OrderReleaseModel_")
        assertEquals(copyOfModels[3].javaClass.simpleName, "OrderReleaseModel_")
        assertEquals(copyOfModels[4].javaClass.simpleName, "OrderReleaseModel_")
        assertEquals(copyOfModels[5].javaClass.simpleName, "OrderReleaseModel_")
        assertEquals(copyOfModels[6].javaClass.simpleName, "DividerModel_")
        assertEquals(copyOfModels[7].javaClass.simpleName, "TotalModel_")
        assertEquals(copyOfModels.size, 8)
    }

    @Test
    fun fourItems_displaysFourItems() {
        controller.setOrderDetails(OrderFactory.buildOneOrderWithItems(4))
        val copyOfModels = controller.adapter.copyOfModels
        assertEquals(copyOfModels[0].javaClass.simpleName, "BuyerModel_")
        assertEquals(copyOfModels[1].javaClass.simpleName, "DividerModel_")
        assertEquals(copyOfModels[2].javaClass.simpleName, "OrderReleaseModel_")
        assertEquals(copyOfModels[3].javaClass.simpleName, "OrderReleaseModel_")
        assertEquals(copyOfModels[4].javaClass.simpleName, "OrderReleaseModel_")
        assertEquals(copyOfModels[5].javaClass.simpleName, "OrderReleaseModel_")
        assertEquals(copyOfModels[6].javaClass.simpleName, "DividerModel_")
        assertEquals(copyOfModels[7].javaClass.simpleName, "TotalModel_")
        assertEquals(copyOfModels.size, 8)
    }
}